Extended internal pattern language of the congruence tactic

mrhaandi opened the pull request:
This PR extends patterns used by the congruence tactic.
Now it allows to represent patterns `?f p1..pn` with `n > 0`.
For example the following failing example is now successful (see zulip [discussion](https://coq.zulipchat.com/#narrow/stream/237977-Coq-users/topic/congruence.20and.20specialization)).

```coq
Axiom p : (nat -> nat) -> nat.
Goal (forall f, p f = f 0) -> forall f, p f = f 0.
Proof. Fail congruence. Admitted.
```

The main observation (also in correspondence with the documentation) is that in the pattern `p ?f = ?f 0` the lhs fully suffices for pattern matching (see example below).

```coq
Definition ap {X Y : Type} (f : X -> Y) (x : X) := f x.

Goal (forall f, p f = ap f 0) -> forall f, p f = ap f 0.
Proof. congruence. Qed.
```

Let us see whether ci will like this one.

<!-- If this is a bug fix, make sure the bug was reported beforehand. -->
Closes  #14651, closes #14662


<!-- If there is a user-visible change in coqc/coqtop/coqchk/coq_makefile behavior and testing is not prohibitively expensive: -->
<!-- (Otherwise, remove this line.) -->
- [x] Added / updated test-suite
- [x] Entry added in the changelog.

herbelin reviewed (approved):
This looks nice and interesting. If I understand correctly, it simulates a primitive form of `ap` which does not need to be mentioned explicitly. So, this is convenient and there shouldn't be an overhead I guess.

Note that I'm not expert in Nelson-Oppen though.

mrhaandi commented:
> This looks nice and interesting. If I understand correctly, it simulates a primitive form of `ap` which does not need to be mentioned explicitly. So, this is convenient and there shouldn't be an overhead I guess.
> 
> Note that I'm not expert in Nelson-Oppen though.

There should be no additional overhead because the `?f t` pattern is still not used for actual matching. However, before `g ?f = ?f t` was discarded entirely because of the restricted pattern language. Now the lhs `g ?f` is considered for matching.

ppedrot commented on plugins/cc/cctac.ml:
```diff
@@ -475,9 +488,10 @@ let cc_tactic depth additionnal_terms =
 let congruence_tac depth l =
-  Tacticals.New.tclTHEN (Tacticals.New.tclREPEAT introf) (cc_tactic depth l)
+  Tacticals.New.tclTHEN
+    (Tacticals.New.tclREPEAT (Tacticals.New.tclFIRST [intro; Tacticals.New.tclTHEN hnf_in_concl intro]))
```
I think it's a bad idea to use `hnf_in_concl` here, even if it corresponds to the inlining of `intro`. It's a very weird reduction tactic that should be phased out.  The way this is written is probably going to be inefficient as well. Since we're going for a change of semantics anyways, I'd rather recommend to do what `intro` does, i.e. match on the goal directly but try to `whd`-reduce if not a product / let-in.

mrhaandi replied:
I was mostly following @herbelin's suggestion https://github.com/coq/coq/issues/14662#issuecomment-880867794.
Since `intro` (as opposed to `introf`) is fast and well behaved, I could (instead of copy-pasting `intro` code) replace in the above `hnf_in_concl` by a custom `whd_in_concl` which is basically `Reductionops.clos_whd_flags CClosure.all env sigma t`.
Trying `intro` twice in the reduction case, the performance hit (for `congruence`, which itself is far more expensive) is negligible.

herbelin replied:
I somehow agree with @ppedrot that "`hnf_concl` is a very weird reduction tactic that should be phased out". I proposed to use it because it is what `intro` does already. A form of `whd_in_concl` which preserves let-ins is probably good.

ppedrot replied:
There is a dedicated flag for that, so `Reductionops.clos_whd_flags CClosure.allnolet` is probably the desired reduction.

mrhaandi replied:
I tried `Reductionops.clos_whd_flags CClosure.allnolet` but it broke a lot of use cases such as proving `Proper (le ==> le) f.` by `congruence`. For this, weak head without let does not produce a good conclusion for a subsequent `intro`.

ppedrot commented on plugins/cc/cctac.ml:
```diff
@@ -43,6 +43,15 @@ let whd env sigma t =
 let whd_delta env sigma t =
   Reductionops.clos_whd_flags CClosure.all env sigma t
 
+let whd_in_concl =
```
Use `Tactics.reduct_in_concl`.

mrhaandi replied:
Ok (I presume `reduct_in_concl ~check:false (Reductionops.whd_all, DEFAULTcast)`).
Since I'm new to Coq's ltac internals, it is not obvious for me which functions are discouraged.

mrhaandi replied:
I'm not sure why `reduct_in_concl` broke `test-suite:base+async` whereas the previous manual `Reductionops.clos_whd_flags CClosure.all` worked fine. I cannot reproduce the error.

herbelin replied:
The async tests (iirc) outputs different things depending on some internal parallelism. If one of `Reductionops.clos_whd_flags CClosure.all` or `reduct_in_concl` is faster than the other, this may change the probabilty that one output comes after the other.

ppedrot commented:
CI green, so this looks ready. @mrhaandi I think you should squash the various commits of this PR though.

mrhaandi commented:
> CI green, so this looks ready. @mrhaandi I think you should squash the various commits of this PR though.

Now squashed and rebased. If ci likes it, then it is good to go. I'd suggest a merge after PR #14657 because (although orthogonal) there could be merge conflicts.

ppedrot commented:
@mrhaandi as expected, this now needs a rebase.

mrhaandi commented:
> @mrhaandi as expected, this now needs a rebase.

Done. ci `signal 9` errors seem due to the main branch.
